### ISSUE TITLE
Sum format returns a space for Kiloseparator

### ISSUE THREAD

**Issue by camponez** [SemanticMediaWiki/SemanticResultFormats#248]
### Setup
- MW version: 1.28.2
- DB (MySQL etc.): 10.2.6-MariaDB-10.2.6+maria~jessie
- PHP version:7.0.16 (fpm-fcgi)
- SMW version: 2.5.2
- SRF version: 2.5.0
- Browsers and versions (if applicable): ALL

### Issue

Sum format returns space for Kiloseparator.

**Steps to reproduce**

Use {{#ask: PARAMETER | format=sum}}

**Expected result:** ...

Either: with the kilo separator defined by the locale (in my case 'a dot')
Or: a standard number format (no kiloseparator and 'a dot' for decimal)

**Observed result:** ...

Kilo separator is a space.

**Comment by mwjames** [comment]
> **Observed result:** ...
>
> Kilo separator is a space.

@jaideraf Could you comment here on the pr-br matter.

As defined in [0], I would expect the decimal as `,` (comma) and kilo
being a ` ` (space) therefore a number like `3,000.2` in English is
displayed in "português do Brasil" as `3 000,2`.

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/commit/e7dea7f0349799fdfee307e44c8f8c184a945e7b

**Comment by camponez** [comment]
@mwjames 
1 - I think it should be `.` for pt-br
2 - I don't see that line on the SW I have installed.

**Comment by camponez** [comment]
When using `table` format all numbers are correctly formatted.

**Comment by mwjames** [comment]
> When using `table` format all numbers are correctly formatted.

I'm confused now, or I misinterpreted the issue description.

I would expect (given the previously outlined rules of comma and
space) that a result for the number 3000.2 (ISO format) is displayed
both in `sum` and `table` format (for user language pt-br) as:

3 000,2

**Comment by camponez** [comment]
>I would expect (given the previously outlined rules of comma and space) that a result for the number 3000.2 (ISO format) is displayed both in `sum` and `table` format (for user language pt-br) as:
> 3 000,2

That's exactly the issue. It only shows `space` for kiloseparator when I use `sum` format. For `table` it works correctly.

Sum format: `3 000,02`
Table format: `3.000,02`

**Comment by camponez** [comment]
Here it works: https://is.gd/bK22hl
Here it doesn't: https://is.gd/YOoXJs

**Comment by mwjames** [comment]
>
> Sum format: `3 000,02`
> Table format: `3.000,02`

It depends on the changes to the SMW-core [0] (available with 2.5.3)
where space is defined as `smw_kiloseparator` for language pt-br and
would make both to be displayed
as:

 `3 000,02`

Consequently it means that the current table format (2.5.2) shows the
wrong notation according to [0].

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/commit/e7dea7f0349799fdfee307e44c8f8c184a945e7b

**Comment by camponez** [comment]
>It depends on the changes to the SMW-core [0] (available with 2.5.3) where space is defined as `smw_kiloseparator` for language pt-br and would make both to be displayed

I have altered my code adding the `sw_kiloseparator` line just like the patch, but using `.` instead. It didn't change anything. Is there any order change that need to be done?

**Comment by mwjames** [comment]
> I have altered my code adding the `sw_kiloseparator` line just like the
> patch, but using `.` instead. It didn't change anything. Is there any order
> change that need to be done?

Now, I understand the issue. The math code uses the standard pt-br
language formatting from MediaWiki as:

```
		// if raw-format ("-") than skip formatNum()
		if ( $outputformat != "-" ) {
			global $wgLang;
			$number = $wgLang->formatNum( $number );
		}
```

This overrides (or better neglects our custom smw_... keys) for the
SMW specific display formatting and just uses the plain user language
formatting.

**Comment by mwjames** [comment]
> This overrides (or better neglects our custom smw_... keys) for the
> SMW specific display formatting and just uses the plain user language

If we wanted to make them behave equally then those lines would need
to be replaced with:

```
 		// if raw-format ("-") than skip formatNum()
 		if ( $outputformat != "-" ) {
-			global $wgLang;
-			$number = $wgLang->formatNum( $number );
+			$dataValue =
\SMW\DataValueFactory::getInstance()->newDataValueByType( '_num' );
+			$number = $dataValue->getLocalizedFormattedNumber( $number );
 		}
```

This relates to #202 (@jongfeli) and would require some integration
tests [0] before it can be green-lit to avoid introducing some form of
regression.

[0] https://github.com/SemanticMediaWiki/SemanticResultFormats/tree/master/tests/phpunit/Integration/JSONScript

**Comment by camponez** [comment]
I will try that. Thank you.

**Comment by kghbln** [comment]
I actually have no longer any idea what the expected kilo separator for "pt-br" is. If a space is wrong it should be changed [here](https://translatewiki.net/wiki/MediaWiki:Smw_kiloseparator/pt-br). I am backing out of this.

**Comment by jaideraf** [comment]
Not sure if the question is already clear...

In Portuguese (and Brazilian Portuguese), decimal is used with a comma [1].
So, "1.5" in English is presented as "1,5" in Portuguese.

"Portuguese uses a dot to separate thousands, eg. 2.367.900 = 2,367,900" [2]

> Here it doesn't: https://is.gd/YOoXJs

This shoud be presented as "336.397,5" or, better yet, "336.397,50" because it is a property for currency.

[1] https://en.wikipedia.org/wiki/Decimal_mark#Hindu.E2.80.93Arabic_numeral_system
[2] http://www.omniglot.com/language/numbers/portuguese.htm

**Comment by kghbln** [comment]
The examples on wikipedia show a [space](https://en.wikipedia.org/wiki/Decimal_mark#Examples_of_use) so I added a "space". @jaideraf Please do me the favour and adjust at translatewiki.net as it fits the "pt-br" needs.

**Comment by mwjames** [comment]
> The examples on wikipedia show a
> [space]

MediaWiki's number formatting also uses spaces but [0, 1] shows a difference in:

- [0] European Portuguese to be `2 000`
- [1] Portuguese to be `2.000`

[0] http://www.unicode.org/cldr/charts/28/verify/numbers/pt_PT.html
[1] http://www.unicode.org/cldr/charts/28/verify/numbers/pt.html

PS: MediaWiki's number formatting uses the CLDR data as basis.

**Comment by jaideraf** [comment]
@kghbln @mwjames I guess it is done now at translatewiki.net: [1][2][3][4]

[1] https://translatewiki.net/wiki/MediaWiki:Smw_kiloseparator/pt-br (.)
[2] https://translatewiki.net/wiki/MediaWiki:Smw_kiloseparator/pt ( )
[3] https://translatewiki.net/wiki/MediaWiki:Smw_decseparator/pt-br (,)
[4] https://translatewiki.net/wiki/MediaWiki:Smw_decseparator/pt (,)

**Comment by kghbln** [comment]
@jaideraf Thanks a lot. One already can have a hard time to figure this out for "de-ch" so "pt" and friends are even more difficult for non-natives.

**Comment by camponez** [comment]
>If we wanted to make them behave equally then those lines would need to be replaced with:
``` // if raw-format ("-") than skip formatNum() 
if ( $outputformat != "-" ) { 
- global $wgLang; 
- $number = $wgLang->formatNum( $number ); 
+ $dataValue = \SMW\DataValueFactory::getInstance()->newDataValueByType( '_num' ); 
+ $number = $dataValue->getLocalizedFormattedNumber( $number ); 
}
```


Yes. That worked for me

**Comment by jaideraf** [comment]
> ... and would require some integration tests...

@mwjames I am trying to understand that JSON test thing to see if I can help a little, but I couldn't make it work. Why, at the wiki, the sum value is 30 030 025,75 but, in the test result, the sum value is 31 517 575? If it's not too wrong, maybe you could adjust this test to make it to work.

```javascript
{
	"description": "Test for sum result format in pt-br lang",
	"setup": [
		{
			"page": "MediaWiki:Smw_kiloseparator/pt-br",
			"namespace": "NS_MEDIAWIKI",
			"contents": "."
		},
		{
			"page": "MediaWiki:Smw_decseparator/pt-br",
			"namespace": "NS_MEDIAWIKI",
			"contents": ","
		},
		{
			"page": "Property:Has_number",
			"namespace": "SMW_NS_PROPERTY",
			"contents": "[[Has type::Number]]"
		},
		{
			"page": "Has_number",
			"namespace":"NS_MAIN",
			"contents": "[[Has number::30000000]] [[Has number::15000]] [[Has number::15000,25]] [[Has number::25,50]] Sum: {{#show:{{PAGENAME}}|?Has number|format=sum}}"
		}
	],
	"tests": [
		{
			"type": "parser",
			"about": "#0 result format for sum in pt-br lang (test for kilo and decimal separators)",
			"subject": "Has_number",
			"assert-output": {
				"to-contain": [
					"30.030.025,75"
				]
			}
		}
	],
	"settings": {
		"wgContLang": "pt-br",
		"wgLang": "pt-br",
		"smwgNamespacesWithSemanticLinks": {
			"NS_MAIN": true,
			"SMW_NS_PROPERTY": true
		}
	},
	"meta": {
		"version": "2",
		"is-incomplete": false,
		"debug": false
	}
}
```

**Comment by mwjames** [comment]
> I am trying to understand that JSON test thing to see if I can help a little, but I couldn't make it work. Why, at the wiki, the sum value is 30 030 025,75 but, in the test result, the sum value is 31 517 575? If it's not too wrong, maybe you could adjust this test to make it to work.

Thanks, I added [0] to clarify possible interferences when designing tests. As for example, I'll wait until the translation arrive and then the modified test should pass.

We are not relying on `"page": "MediaWiki:Smw_kiloseparator/pt-br",` because MW's message system reads from the `JSON` files and I have found no good method to read from the DB (or page) during testing therefore we have to wait until the translation from tnw arrives.

```
{
	"description": "Test for math/sum result format in pt-br lang",
	"setup": [
		{
			"page": "Has number",
			"namespace": "SMW_NS_PROPERTY",
			"contents": "[[Has type::Number]]"
		},
		{
			"page": "Example/Math/1",
			"contents": "[[Has number::30000000]] [[Has number::15000]] [[Has number::15000,25]] [[Has number::25,50]] "
		},
		{
			"page": "Example/Math/Q.1",
			"contents": "{{#show: Example/Math/1 |?Has number|format=sum }}"
		},
		{
			"page": "Example/Math/Q.2",
			"contents": "{{#show: Example/Math/1 |?Has number|format=max }}"
		},
		{
			"page": "Example/Math/Q.3",
			"contents": "{{#show: Example/Math/1 |?Has number|format=min }}"
		},
		{
			"page": "Example/Math/Q.4",
			"contents": "{{#show: Example/Math/1 |?Has number|format=product }}"
		},
		{
			"page": "Example/Math/Q.5",
			"contents": "{{#show: Example/Math/1 |?Has number|format=average }}"
		},
		{
			"page": "Example/Math/Q.6",
			"contents": "{{#show: Example/Math/1 |?Has number|format=median }}"
		}
	],
	"tests": [
		{
			"type": "parser",
			"about": "#0 format=sum (kilo and decimal separators)",
			"subject": "Example/Math/Q.1",
			"assert-output": {
				"to-contain": [
					"30.030.025,75"
				]
			}
		},
		{
			"type": "parser",
			"about": "#1 format=max (kilo and decimal separators)",
			"subject": "Example/Math/Q.2",
			"assert-output": {
				"to-contain": [
					"30.000.000"
				]
			}
		},
		{
			"type": "parser",
			"about": "#2 format=min (kilo and decimal separators)",
			"subject": "Example/Math/Q.3",
			"assert-output": {
				"to-contain": [
					"25,5"
				]
			}
		},
		{
			"type": "parser",
			"about": "#3 format=product (kilo and decimal separators)",
			"subject": "Example/Math/Q.4",
			"assert-output": {
				"to-contain": [
					"1,721279e+17"
				]
			}
		},
		{
			"type": "parser",
			"about": "#4 format=average (kilo and decimal separators)",
			"subject": "Example/Math/Q.5",
			"assert-output": {
				"to-contain": [
					"7.507.506,438"
				]
			}
		},
		{
			"type": "parser",
			"about": "#5 format=median (kilo and decimal separators)",
			"subject": "Example/Math/Q.6",
			"assert-output": {
				"to-contain": [
					"15.000,125"
				]
			}
		}
	],
	"settings": {
		"wgContLang": "pt-br",
		"wgLang": "pt-br",
		"smwgNamespacesWithSemanticLinks": {
			"NS_MAIN": true,
			"SMW_NS_PROPERTY": true
		}
	},
	"meta": {
		"version": "2",
		"is-incomplete": false,
		"debug": false
	}
}
```

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/commit/d0f8494a27231be15ba8bbf29cb6824b2147f0a9

**Comment by jaideraf** [comment]
Oh, much better, thanks. :clap:

**Comment by mwjames** [comment]
Addressed with #250.
